Merge Dev to Prod to introduce starboard functionality to DrBot

DrHooBs opened the pull request.

InimicalPart committed:
Merge pull request #17 from Incoverse/prod

DrHooBs committed:
✨ Add StreamMessage Command

Add a stream message command to facilitate the sending of live announcements for the streamer

DrHooBs committed:
Update src/commands/command-lib/mod/streammessage.cmdlib.ts

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
🎨 Improve Format

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄Make reply better

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄Fix response

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
💄 More response changes

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
Update Copyright

Co-authored-by: Inimi <me@inimicalpart.com>

DrHooBs committed:
🎨 Add configurable templating

Add configurable templating for the stream message command. Also rename file and command to stream-message. This should be the final fix for this PR

DrHooBs committed:
✨ STARBOARD :D

Add a starboard

DrHooBs commented:
@InimicalPart Let me know if you want me to open a PR without the Stream Message command...
I forgot it was just chilling on dev. It is functional tho :)

InimicalPart commented on src/commands/command-lib/mod/stream-message.cmdlib.ts:
Update this to ``stream-message.cmdlib.d-ts`` so that it's disabled, I don't think we need this as of right now

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,92 @@
+                text: `${reaction.message.id} • DrBot's Starboard`
+            })
+
+        if (reaction.emoji.name == "MrSzanto") {
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,92 @@
+
+        if (reaction.emoji.name == "MrSzanto") {
+            let guild = reaction.message.guild as Guild;
+            let starboardChannel = guild.channels.cache.find(channel => channel.name === "starboard");
```
This is not gonna work due to the fact that the channels in DrVem's Community Centre start with "``»・``", a better way would be to add a regex check, for example:
```js
/star(-){0,1}board/gi
```
which would trigger for channels such as: ``»・starboard``, ``starboard``, ``»・star-board``, and ``star-board``

```suggestion
            let starboardChannel = guild.channels.cache.find(channel => /star(-){0,1}board/gi.test(channel.name)); // Find a channel that has "starboard" in the name (varying formats)
```

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,92 @@
+            if (count >= 1) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
String matching for footers like it is done here is very bad practice, and can cause problems if things change Discord side, I'd highly recommend using another method of checking

This isn't a crucial thing to change, but I highly recommend doing so

DrHooBs replied:
I think for now its best to leave it like this. I am happy to find a new solution **if** discord changes the format of embeds

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,92 @@
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
+                if (existing) {
+                        console.log("Message already starred");
+                        if (count > existing.content.match(/<:MrSzanto:1277693330280550450> (\d+)/)[1]) {
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart replied:
String matching like it is done here is very bad practice, and can cause problems in the future, I'd highly recommend using another method.

This isn't a crucial thing to change, but I highly recommend doing so

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
Did the name get messed up for this file?

InimicalPart replied:
This file name is a bit long, can we try to shorten it down?

DrHooBs replied:
I think it would be best to leave the file as onReactionRemoveUpdateStarboard.evt.ts as it clearly details what the file is for. I have removed the typo

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2024 Inimi | DrHooBs | Incoverse
```
```suggestion
 * Copyright (c) 2025 Inimi | DrHooBs | Incoverse
```

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2024 Inimi | DrHooBs | Incoverse
```
```suggestion
 * Copyright (c) 2025 Inimi | DrHooBs | Incoverse
```

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
This file name is a bit long, can we try to shorten it down?

DrHooBs replied:
Like i said on the other one I think its better to have a long name and it be clear what the file is for.

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+        if (reaction.emoji.name == "MrSzanto") {
+            
+            let guild = reaction.message.guild as Guild;
+            let starboardChannel = guild.channels.cache.find(channel => channel.name === "starboard");
```
This is not gonna work due to the fact that the channels in DrVem's Community Centre start with "``»・``", a better way would be to add a regex check, for example:
```js
/star(-){0,1}board/gi
```
which would trigger for channels such as: ``»・starboard``, ``starboard``, ``»・star-board``, and ``star-board``

```suggestion
            let starboardChannel = guild.channels.cache.find(channel => /star(-){0,1}board/gi.test(channel.name)); // Find a channel that has "starboard" in the name (varying formats)
```

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+            })
+
+
+        if (reaction.emoji.name == "MrSzanto") {
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+            if (count >= 1) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
String matching for footers like it is done here is very bad practice, and can cause problems if things change Discord side, I'd highly recommend using another method of checking

This isn't a crucial thing to change, but I highly recommend doing so

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+                else {
+                
+                    await (starboardChannel as TextChannel).send({
+                        content: `<:MrSzanto:1277693330280550450> ${reaction.count}`,
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+                        } else */
+                        if (count != existing.content.match(/<:MrSzanto:1277693330280550450> (\d+)/)[1]) {
+                            existing.edit({
+                                content: `<:MrSzanto:1277693330280550450> ${count}`,
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+                        /*if (count == 0) {
+                            existing.delete();
+                        } else */
+                        if (count != existing.content.match(/<:MrSzanto:1277693330280550450> (\d+)/)[1]) {
```
Maybe better to specify it in a ``config.jsonc`` option?

InimicalPart commented on src/events/onReactionRemovepdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+            } else if (count == 0) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
String matching for footers like it is done here is very bad practice, and can cause problems if things change Discord side, I'd highly recommend using another method of checking

This isn't a crucial thing to change, but I highly recommend doing so

InimicalPart commented on src/interfaces/appInterface.d.ts:
Can be ignored due to the stream message command being disabled.

InimicalPart commented on config.template.jsonc:
Can be ignored due to the stream message command being disabled.

InimicalPart commented:
I'm also gonna wait to merge this until I can talk to DrVem and see if he likes this idea (since merging with prod will load it into DrBot automatically)

InimicalPart commented:
Generally, I'd also highly ask of you to not hardcode "DrBot" when it is shown publicly. Instead, get the bot's display name and use that, i think you can access the name in the ``client.user`` (``Client#User``) object

DrHooBs committed:
Add jsonc config and regex for channel matching

DrHooBs commented:
I've added jsonc configuration options as requested and changed the footer to get the bots display name. I don't think the file names should be shortened as they follow the `on event action` format of all of the files in that directory. The file names clearly denote what the file is for.

DrHooBs committed:
Add Regex to exclude staff channels

DrHooBs commented:
Forgot to add regex to prevent starboarding messages in staff channels.

InimicalPart commented on config.template.jsonc:
```diff
@@ -240,6 +236,23 @@
+    "cancelled": "[updates-ping], the next stream is cancelled because [reason]. See you all next time!", // @Live Updates, the next stream is cancelled because I'm being sick. See you all next time!"
+    "streaming": "[notifs-ping], we are currently live and playing [game]! Come join us over on Twitch! https://twitch.tv/example", // @Live Notifications, we are currently live and playing Valheim! Come join us over on Twitch! https://twitch.tv/drvem
+  },
+  //? These settings are for the Starboard functionality. Set all of these to null to disable the feature. Like this -> ""
```
That isn't null but whatever, writing literally
```jsonc
{
    // ...
    
    "starboardNumber": null
}
```

is making it null, maybe fix before merging?

DrHooBs replied:
That was pretty bad thanks for picking up on that

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+    
+
+    public async runEvent(reaction, user, client: Client) {
+        if (global.app.config.starboardEmojiID === "" || global.app.config.starboardEmojiSTR === "" || global.app.config.starboardNumber === 0) {
```
```suggestion
        if (!global.app.config.starboardEmojiID || !global.app.config.starboardEmojiSTR || !global.app.config.starboardNumber) {
```

InimicalPart commented on config.template.jsonc:
```diff
@@ -240,6 +236,23 @@
+    "cancelled": "[updates-ping], the next stream is cancelled because [reason]. See you all next time!", // @Live Updates, the next stream is cancelled because I'm being sick. See you all next time!"
+    "streaming": "[notifs-ping], we are currently live and playing [game]! Come join us over on Twitch! https://twitch.tv/example", // @Live Notifications, we are currently live and playing Valheim! Come join us over on Twitch! https://twitch.tv/drvem
+  },
+  //? These settings are for the Starboard functionality. Set all of these to null to disable the feature. Like this -> ""
```
```suggestion
  //? These settings are for the Starboard functionality. Set any of these to null to disable the feature.

```

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+                return;
+            }
+        }
+        if (reaction.message.channel.name.match(/^(staff|mod|admin)([-_][a-z0-9]+)*$/i)) {
```
To be honest, I'd say the regex here is really extensive, just checking for "staff" "mod" or "admin" is usually enough, you don't need the rest.

This also won't work since you're starting the regex with ^, meaning start of the line, but as I mentioned in the previous review, the channels don't start on stuff like "staff", it starts with ``»・``, meaning your regex would fail.


```suggestion
        if (reaction.message.channel.name.match(/(staff|mod|admin)/i)) {
```

DrHooBs replied:
Yeah, sorry about that. I'm not too experienced with regex as I have only used it for one project at school lmfao.

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+                value: `[Jump to message](${reaction.message.url})`
+            })
+            .setFooter({
+                text: `${reaction.message.id} • ${client.user.username}'s Starboard`
```
```suggestion
                text: `${reaction.message.id} • ${client.user.displayName}'${client.user.displayName.toLowerCase().endsWith("s") ? "" : "s"} Starboard`
```

While ``.username`` works, ``.displayName`` is better practice in this case. The other change makes it so that if the name ends with "s", it won't add the "s" after the apostrophe (personal preference and it is what's used across other commands / events. it's better to say consistant)

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+
+        let embed = new EmbedBuilder()
+            .setAuthor({
+                name: reaction.message.author.username,
```
I'm only gonna comment on this one but this applies for all of them, wouldn't ``.displayName`` be better here too?

DrHooBs replied:
Fixed. 👍

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+            })
+
+
+        if (reaction.emoji.name == "MrSzanto") {
```
> Maybe better to specify it in a ``config.jsonc`` option?
> *__From previous review__*

DrHooBs replied:
That's really odd my local version was using json config. I will update this.

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+            let guild = reaction.message.guild as Guild;
+            let starboardChannel = guild.channels.cache.find(channel => /star(-){0,1}board/gi.test(channel.name)); // Find a channel that has "starboard" in the name (varying formats)
+            let count = reaction.count;
+            if (reaction.message.channel.id == starboardChannel.id) { return }
```
```suggestion
            if (reaction.message.channel.id == starboardChannel.id) return;
```

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+            if (count >= global.app.config.starboardNumber) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
Please see https://github.com/Incoverse/DrBot/pull/20#issuecomment-2816825735

InimicalPart commented on src/events/onReactionRemoveUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,109 @@
+            } else if (count == 0) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
Please see https://github.com/Incoverse/DrBot/pull/20#issuecomment-2816825735

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+    
+
+    public async runEvent(reaction, user, client: Client) {
+        if (global.app.config.starboardEmojiID === "" || global.app.config.starboardEmojiSTR === "" || global.app.config.starboardNumber === 0) {
```
```suggestion
        if (!global.app.config.starboardEmojiID || !global.app.config.starboardEmojiSTR || !global.app.config.starboardNumber) {
```

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+        }
+
+    
+        if (reaction.message.channel.name.match(/^(staff|mod|admin)([-_][a-z0-9]+)*$/i)) {
```
To be honest, I'd say the regex here is really extensive, just checking for "staff" "mod" or "admin" is usually enough, you don't need the rest.

This also won't work since you're starting the regex with ^, meaning start of the line, but as I mentioned in the previous review, the channels don't start on stuff like "staff", it starts with ``»・``, meaning your regex would fail.


```suggestion
        if (reaction.message.channel.name.match(/(staff|mod|admin)/i)) {
```

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+
+        let embed = new EmbedBuilder()
+            .setAuthor({
+                name: reaction.message.author.username,
```
I'm only gonna comment on this one but this applies for all of them, wouldn't ``.displayName`` be better here too?

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+                value: `[Jump to message](${reaction.message.url})`
+            })
+            .setFooter({
+                text: `${reaction.message.id} • ${client.user.username}'s Starboard`
```
```suggestion
                text: `${reaction.message.id} • ${client.user.displayName}'${client.user.displayName.toLowerCase().endsWith("s") ? "" : "s"} Starboard`
```

While ``.username`` works, ``.displayName`` is better practice in this case. The other change makes it so that if the name ends with "s", it won't add the "s" after the apostrophe (personal preference and it is what's used across other commands / events. it's better to say consistant)

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+            if (count >= global.app.config.starboardNumber) {
+                //Check if message is already starred
+                let msgs = await (starboardChannel as TextChannel).messages.fetch();
+                const existing = msgs.find(msg => msg.embeds[0]?.footer?.text == `${reaction.message.id} • DrBot's Starboard`);
```
Please see https://github.com/Incoverse/DrBot/pull/20#issuecomment-2816825735

InimicalPart commented on src/events/onReactionAddUpdateStarboard.evt.ts:
```diff
@@ -0,0 +1,104 @@
+        if (reaction.emoji.name == global.app.config.starboardEmojiSTR) {
+            let guild = reaction.message.guild as Guild;
+            let starboardChannel = guild.channels.cache.find(channel => /star(-){0,1}board/gi.test(channel.name)); // Find a channel that has "starboard" in the name (varying formats)
+            if (reaction.message.channel.id == starboardChannel.id) { return }
```
```suggestion
            if (reaction.message.channel.id == starboardChannel.id) return;
```

InimicalPart reviewed (changes requested):
You've not done some of the changes requested from the previous review. Please look through the comments and double-check before you commit 🙏

InimicalPart commented:
> I'm also gonna wait to merge this until I can talk to DrVem and see if he likes this idea (since merging with prod will load it into DrBot automatically)

Also please take note of this.

DrHooBs committed:
Change footer to remove hardcoding and other fixes

Changed footer to be dynamic
Update ORUS to use app config
"" to null in config template
embed changed to use display name

InimicalPart reviewed (approved):
Seems okay, I'm gonna make some changes to the config names later, but I think this works for now.

Great work Hoobs!
